src/hotspot/share/gc/shenandoah/shenandoahFreeSet.hpp

```diff
@@ -29,30 +29,136 @@
+  Mutator,
+  Collector,
+  OldCollector,
+  NumFreeSets
+};
+
+class ShenandoahSetsOfFree {
+
 private:
-  ShenandoahHeap* const _heap;
-  CHeapBitMap _mutator_free_bitmap;
-  CHeapBitMap _collector_free_bitmap;
-  size_t _max;
+  size_t _max;                  // The maximum number of heap regions
+  ShenandoahFreeSet* _free_set;
+  size_t _region_size_bytes;
+  FreeMemoryType* _membership;
+  size_t _left_mosts[NumFreeSets];
+  size_t _right_mosts[NumFreeSets];
+  size_t _left_mosts_empty[NumFreeSets];
+  size_t _right_mosts_empty[NumFreeSets];
+  size_t _capacity_of[NumFreeSets];
+  size_t _used_by[NumFreeSets];
+  bool _left_to_right_bias[NumFreeSets];
+  size_t _region_counts[NumFreeSets];
+
+  inline void shrink_bounds_if_touched(FreeMemoryType set, size_t idx);
+  inline void expand_bounds_maybe(FreeMemoryType set, size_t idx, size_t capacity);
+
+  // Restore all state variables to initial default state.
+  void clear_internal();
 
-  // Left-most and right-most region indexes. There are no free regions outside
-  // of [left-most; right-most] index intervals
-  size_t _mutator_leftmost, _mutator_rightmost;
-  size_t _collector_leftmost, _collector_rightmost;
+public:
+  ShenandoahSetsOfFree(size_t max_regions, ShenandoahFreeSet* free_set);
+  ~ShenandoahSetsOfFree();
+
+  // Make all regions NotFree and reset all bounds
+  void clear_all();
+
+  // Remove or retire region idx from all free sets.  Requires that idx is in a free set.  This does not affect capacity.
+  void remove_from_free_sets(size_t idx);
+
+  // Place region idx into free set which_set.  Requires that idx is currently NotFree.
+  void make_free(size_t idx, FreeMemoryType which_set, size_t region_capacity);
+
+  // Place region idx into free set new_set.  Requires that idx is currently not NotFRee.
+  void move_to_set(size_t idx, FreeMemoryType new_set, size_t region_capacity);
+
+  // Returns the FreeMemoryType affiliation of region idx, or NotFree if this region is not currently free.  This does
+  // not enforce that free_set membership implies allocation capacity.
+  inline FreeMemoryType membership(size_t idx) const;
+
+  // Returns true iff region idx is in the test_set free_set.  Before returning true, asserts that the free
+  // set is not empty.  Requires that test_set != NotFree or NumFreeSets.
+  inline bool in_free_set(size_t idx, FreeMemoryType which_set) const;
+
+  // The following four methods return the left-most and right-most bounds on ranges of regions representing
+  // the requested set.  The _empty variants represent bounds on the range that holds completely empty
+  // regions, which are required for humongous allocations and desired for "very large" allocations.  A
+  // return value of -1 from left_most() or left_most_empty() denotes that the corresponding set is empty.
+  // In other words:
+  //   if the requested which_set is empty:
+  //     left_most() and left_most_empty() return _max, right_most() and right_most_empty() return 0
```

Comment:
I think you an elide the "_" underscore in left_most (leftmost is an English word). It also makes `leftmost_empty` easier to read.

Would be great if you can make that global substitution in your comments etc. as well before checking in. (Or make it all have the underscore consistently in comments too if you prefer the variant with the underscore.)

Reply:
Thanks.  Done.